tmt/utils/signals.py

```diff
@@ -0,0 +1,149 @@
+"""
+Signal handling in tmt.
+
+Provides custom signal handler for ``SIGINT`` (aka ``Ctrl+C``) and
+``SIGTERM``, and delayed actions for these signals.
+
+Some of the factors tmt need to take into account when dealing with
+signals:
+
+* tmt runs a lot of things in dedicated threads,
+* sometimes tmt cannot let signals interrupt whatever is being done at
+  the moment, namely saving API responses from provisioning services,
+* signals in Python are always executed in the context of the main
+  thread,
+* signals are most often delivered to the main thread because that is
+  what ``Ctrl+C`` will deliver to,
+* masking signals in a thread will not affect other threads.
+
+To support uninterruptible blocks, signal handler and
+:py:class:`PreventSignals` use events to record whether a signal was
+delivered and needs handling:
+
+* if signals are not blocked, signal handler would raise the
+  :py:exc:`KeyboardInterrupt` exception right away,
+* if signals are blocked, signal handler would record its delivery, and
+  :py:class:`PreventSignals` would raise the :py:exc:`KeyboardInterrupt`
+  when leaving the uninterruptible block.
+"""
+
+import contextlib
+import signal
+import textwrap
+import threading
+from types import FrameType
+from typing import NoReturn, Optional
+
+import tmt.log
+
+#: All changes to :py:data:`_INTERRUPT_MASKED` and
+#: :py:data:`_INTERRUPT_PENDING` must be performed while holding this
+#: lock. It prevents inconsistencies between handlers and
+#: uninterruptible blocks.
+_INTERRUPT_LOCK = threading.Lock()
+
+#: When set, it is not allowed to interrupt what tmt is doing.
+_INTERRUPT_MASKED = threading.Event()
+
+#: When set, interrupt was delivered to tmt, and tmt should react to it.
+_INTERRUPT_PENDING = threading.Event()
+
+
+def _quit_tmt(logger: tmt.log.Logger, repeated: bool = False) -> NoReturn:
+    """
+    Send tmt on the path of quitting by raising an exception.
+    """
+
+    if repeated:
+        logger.warning(
+            textwrap.dedent(
+                """
+              Repeated interruption requested.
```

Comment:
Align under the trip quotes? Similar below.

Reply:
Addressed in 677e4f05